pkg/provisioner/ironic/ironic_test.go

```diff
@@ -313,3 +374,129 @@ func TestGetUpdateOptsForNodeDell(t *testing.T) {
 		})
 	}
 }
+
+func makeHost() *metal3v1alpha1.BareMetalHost {
+	rotational := true
+
+	return &metal3v1alpha1.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "myhost",
+			Namespace: "myns",
+			UID:       "27720611-e5d1-45d3-ba3a-222dcfaa4ca2",
+		},
+		Spec: metal3v1alpha1.BareMetalHostSpec{
+			Image: &metal3v1alpha1.Image{
+				URL: "not-empty",
+			},
+			Online:          true,
+			HardwareProfile: "libvirt",
+			RootDeviceHints: &metal3v1alpha1.RootDeviceHints{
+				DeviceName:         "userd_devicename",
+				HCTL:               "1:2:3:4",
+				Model:              "userd_model",
+				Vendor:             "userd_vendor",
+				SerialNumber:       "userd_serial",
+				MinSizeGigabytes:   40,
+				WWN:                "userd_wwn",
+				WWNWithExtension:   "userd_with_extension",
+				WWNVendorExtension: "userd_vendor_extension",
+				Rotational:         &rotational,
+			},
+		},
+		Status: metal3v1alpha1.BareMetalHostStatus{
+			Provisioning: metal3v1alpha1.ProvisionStatus{
+				ID: "provisioning-id",
+				// Place the hints in the status field to pretend the
+				// controller has already reconciled partially.
+				RootDeviceHints: &metal3v1alpha1.RootDeviceHints{
+					DeviceName:         "userd_devicename",
+					HCTL:               "1:2:3:4",
+					Model:              "userd_model",
+					Vendor:             "userd_vendor",
+					SerialNumber:       "userd_serial",
+					MinSizeGigabytes:   40,
+					WWN:                "userd_wwn",
+					WWNWithExtension:   "userd_with_extension",
+					WWNVendorExtension: "userd_vendor_extension",
+					Rotational:         &rotational,
+				},
+			},
+			HardwareProfile: "libvirt",
+		},
+	}
+}
+
+func newMockServer(port int) *mockServer {
+	return &mockServer{
+		port: strconv.Itoa(port),
+	}
+}
+
+type mockServer struct {
+	port             string
+	requests         string
+	server           *httptest.Server
+	drivers          string
+	hasInternalError bool
+}
+
+func (m *mockServer) setInternalError() *mockServer {
+	m.hasInternalError = true
+
+	return m
+}
+
+func (m *mockServer) addDrivers() *mockServer {
+	m.drivers = `
+	{
+		"drivers": [{
+			"hosts": [
+			  "master-2.ostest.test.metalkube.org"
+			],
+			"links": [
+			  {
+				"href": "http://[fd00:1101::3]:6385/v1/drivers/fake-hardware",
+				"rel": "self"
+			  },
+			  {
+				"href": "http://[fd00:1101::3]:6385/drivers/fake-hardware",
+				"rel": "bookmark"
+			  }
+			],
+			"name": "fake-hardware"
+		}]
+	}
+	`
+	return m
+}
+
+func (m *mockServer) start() *mockServer {
+	listener, err := net.Listen("tcp", "127.0.0.1:"+m.port)
+	if err != nil {
+		panic(err)
+	}
+
+	m.server = httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		m.requests += r.Host + r.RequestURI + ";"
```

Comment:
It was done by design in reality to improve test readability / handling: with this testing technique events are accumulated in a string (using a separator), so that it would be easier to setup the expected string (and to read in case of failure, also the sequence is naturally represented). It's an approach that works well especially when there a few events.